### PR TITLE
BF: Fixing everything needed for demo script

### PR DESCRIPTION
**Pull request by bpoldrack**
This is based on cherry-picked commit from `docs` branch as well as on `master`.
Purpose is to collect all necessary fixes to get the demo script(s) running.

### CODE REVIEW

**Comment by hanke** [comment]
@yarikoptic @bpoldrack I think the time has come for finally naming what is now `Dataset.remember_state()`:

Options are (please add yours):

- commit()
- save()/saveas() (with an anticipated "load" counterpart
- record()
- remember()

I personally, suggested 'remember/record'. I dislike 'commit' for clashing with git commit. I gotten to like "save()".

**Comment by bpoldrack** [comment]
+1 save/load

It's a known concept to everyone, that provides a basis understanding of what this is about, while these are still terms, that are used in a very general way and therefore are not too strictly defined already. You don't need to know or understand git to get the idea of saving and loading a state.
The other way around: If you understand git, you will be able to deal with these terms and anticipate them, I think.

**Comment by yarikoptic** [comment]
I guess that 'save/load' would indeed be at least easier to comprehend. Just to re-repeat myself though, I do not think that coming up with new names just not to clash with git is imho advantageous.

FWIW, here is how --help summary of commands look at this moment (as of this PR  0.2-289-ge3436a6):
```
Commands for dataset operations

  - create:  Create a new dataset
  - install:  Install a dataset component or entire datasets
  - publish:  publish a handle
  - uninstall:  Uninstall a dataset component or entire datasets
  - move:  Move a handle
  - update:  Update a dataset from a sibling
  - create-publication-target-sshwebserver:  Create a dataset on a web server via SSH, that may then serve as
  - add-sibling:  Adds a sibling to a dataset
  - modify-subhandle-urls:  Modify the URLs of subdatasets of a dataset
  - unlock:  Proxying annex unlock
  - make-memory-engram:  see Dataset.remember_state()

Miscellaneous commands

  - test:  Run internal DataLad (unit)tests
  - crawl:  Crawl online resource to create or update a handle
  - pull:  Get changes from a remote repository
  - push:  Push changes to a remote repository
  - ls:  Magical helper to list content of various things (ATM only S3 buckets and
        datasets)
  - clean:  Clean up after DataLad (e.g. extracted temporary archives etc)
  - add-archive-content:  Add content of an archive under git annex control
  - download-url:  Download a content from a URL using DataLad's downloader
  - create-test-dataset:  Create test (meta-)dataset
```
to me, as a git experienced user, or a newbee all those siblings and handles are alien things. 
For that matter, somewhat animate 'sibling' is IMHO doesn't go well with 'dataset'.  We do use appropriate super- sub- (not parent, child), so 'remote' (yes, colliding with git `remote` and actually implemented as a git remote) would have made imho more sense.  If `save` is to do just a bunch of commits, I would have also preferred `commit`.
My 1 c

File: datalad/distribution/update.py [hunk 1]
```diff
@@ -127,27 +132,35 @@ def __call__(name=None, dataset=None,
...
+                    # check for tracking branch's remote:
+                    try:
+                        std_out, std_err = \
+                            repo._git_custom_command('',
+                                                     ["git", "config", "--get",
+                             "branch.{active_branch}.remote".format(
+                                 active_branch=repo.get_active_branch())])
+                    except CommandError as e:
+                        if e.code == 1 and e.stdout == "":
+                            std_out = None
```
Comment (yarikoptic) [inline datalad/distribution/update.py]: just to be safe, `std_out, std_err = None, None`  may be name consistently with/in other code below where it is `out, err`

File: datalad/interface/make_memory_engram.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+from datalad.distribution.dataset import datasetmethod
+from datalad.utils import getpwd
+
+from .base import Interface
+
+lgr = logging.getLogger('datalad.interface.commit')
+
+
+class MakeMemoryEngram(Interface):
+    """see Dataset.remember_state()."""
```
Comment (yarikoptic) [inline datalad/interface/make_memory_engram.py]: I wonder if `@borrow_doc` could be used. Remember that this line is used for generating overall --help so it is important to have it 'proper'

Comment (hanke) [inline datalad/interface/make_memory_engram.py]: I'd vote for handwritten docs.

**Comment by hanke** [comment]
So to conclude this: We go for `save()`. Reason for not using commit is that `save` combines particular invocations of `add`, `commit`, and possibly `tag` into a single command. This breaks the Git logic of these things being distinct steps.

**Comment by hanke** [comment]
@bpoldrack What is the state here? It says WiP. It has also accumulated conflict(s). Who is fixing?

**Comment by bpoldrack** [comment]
I'm fixing it. Despite that, I wanted to see whether we come up with additional needs for that demo script. It reads "first usecase" in its comments ...

File: datalad/distribution/dataset.py [hunk 3]
```diff
@@ -222,6 +222,8 @@ def get_dataset_handles(self, pattern=None, fulfilled=None, absolute=False,
 #        raise NotImplementedError("TODO")
 
     # TODO maybe needs to get its own interface
+    # TODO: issue: auto_add_changes isn't that easy. consider an annex repo with
+    #       with changes to a file in git. We would currently call 'annex add .'
```
Comment (hanke) [inline datalad/distribution/dataset.py]: FTR: For an annex repo this is IMHO:

`git annex add .`

Deals with everything that is not already handled by Git directly. Followed by:

`git add .`

Does the rest.

File: datalad/distribution/dataset.py [hunk 4]
```diff
@@ -222,6 +222,8 @@ def get_dataset_handles(self, pattern=None, fulfilled=None, absolute=False,
...
     def remember_state(self, message, auto_add_changes=True, version=None):
```
Comment (hanke) [inline datalad/distribution/dataset.py]: We should now unify to be `save()`

File: datalad/distribution/dataset.py [hunk 5]
```diff
@@ -222,6 +222,8 @@ def get_dataset_handles(self, pattern=None, fulfilled=None, absolute=False,
 #        raise NotImplementedError("TODO")
 
     # TODO maybe needs to get its own interface
```
Comment (hanke) [inline datalad/distribution/dataset.py]: Remove. It got it's own.

File: datalad/distribution/update.py [hunk 6]
```diff
@@ -129,33 +135,43 @@ def __call__(name=None, dataset=None,
...
+                    # TODO: what does failing here look like?
+                else:
+                    # we have no remote given, therefore
+                    # check for tracking branch's remote:
+                    try:
+                        std_out, std_err = \
+                            repo._git_custom_command('',
+                                                     ["git", "config", "--get",
+                             "branch.{active_branch}.remote".format(
+                                 active_branch=repo.get_active_branch())])
```
Comment (hanke) [inline datalad/distribution/update.py]: Any advantage over using the config reader of `repo` directly?

File: datalad/distribution/update.py [hunk 7]
```diff
@@ -129,33 +135,43 @@ def __call__(name=None, dataset=None,
...
 
             # merge:
             if merge:
                 lgr.info("Applying changes from tracking branch...")
                 # TODO: Adapt.
-                # TODO: Rethink default remote/tracking branch. See above. We need a "tracking remote" but custom refspec to fetch from that remote
+                # TODO: Rethink default remote/tracking branch. See above.
+                # We need a "tracking remote" but custom refspec to fetch from
+                # that remote
                 cmd_list = ["git", "pull"]
```
Comment (hanke) [inline datalad/distribution/update.py]: Shouldn't this use `GitRepo.pull()`?

File: datalad/distribution/update.py [hunk 8]
```diff
@@ -167,12 +183,13 @@ def __call__(name=None, dataset=None,
...
-                lgr.info(out)
+                std_out, std_err = repo._git_custom_command('', cmd_list)
+                lgr.info(std_out)
                 if knows_annex(repo.path):
                     # annex-apply:
                     lgr.info("Updating annex ...")
-                    out, err = repo._git_custom_command('', ["git", "annex", "merge"])
-                    lgr.info(out)
+                    std_out, std_err = repo._git_custom_command(
+                        '', ["git", "annex", "merge"])
```
Comment (hanke) [inline datalad/distribution/update.py]: We should move this into a `AnnexRepo.merge()` method to have things in one place.

File: datalad/interface/save.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+from datalad.distribution.dataset import datasetmethod
+from datalad.utils import getpwd
+
+from .base import Interface
+
+lgr = logging.getLogger('datalad.interface.commit')
+
+
+class Save(Interface):
+    """Save the current state of a dataset."""
```
Comment (hanke) [inline datalad/interface/save.py]: This causes sphinx pain. Always do:

````
"""Some summary

Ideally some more blurb.
"""
````

File: datalad/interface/save.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        if ds is not None and not isinstance(ds, Dataset):
+            ds = Dataset(ds)
+
+        if ds is None:
+            # try CWD:
+            dspath = GitRepo.get_toppath(getpwd())
+            if not dspath:
+                raise InsufficientArgumentsError("No dataset found")
+            ds = Dataset(dspath)
+
```
Comment (hanke) [inline datalad/interface/save.py]: The entire start of the function should be a common pattern by now. Time for a function?

File: datalad/interface/save.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        ds.remember_state(message=message,
+                          auto_add_changes=auto_add_changes,
+                          version=version_tag)
+
+        # TODO: Should remember_state return the commit SHA?
```
Comment (hanke) [inline datalad/interface/save.py]: Why not? Better than nothing.

File: datalad/interface/unlock.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+from datalad.distribution.dataset import datasetmethod
+from datalad.utils import getpwd
+
+from .base import Interface
+
+lgr = logging.getLogger('datalad.interface.unlock')
+
+
+class Unlock(Interface):
+    """Unlock file(s) of a dataset"""
```
Comment (hanke) [inline datalad/interface/unlock.py]: Sphinx trouble. See above.

File: datalad/interface/unlock.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+            ds = Dataset(dspath)
+
+        assert ds
+        assert ds.repo
+
+        if not isinstance(ds.repo, AnnexRepo):
+            # TODO: Introduce NoAnnexError
+            raise ValueError("No annex found in dataset (%s)." % ds.path)
+
+        # TODO: AnnexRepo().unlock() with proper return value
```
Comment (hanke) [inline datalad/interface/unlock.py]: +1 from me

**Comment by hanke** [comment]
FTR: Travis failure, at least in part, to the new and much stricter requirements on docstring syntax correctness -- see in-line comments above.

File: datalad/distribution/update.py (hunk 6 above)
Comment (bpoldrack) [inline datalad/distribution/update.py]: Nope. Just remained from initial implementation, which purposefully didn't use gitpython.

File: datalad/interface/save.py (hunk 10 above)
Comment (bpoldrack) [inline datalad/interface/save.py]: Yes.

File: datalad/interface/save.py (hunk 11 above)
Comment (bpoldrack) [inline datalad/interface/save.py]: Agree, just wondered whether someone would come up with something better.

File: datalad/distribution/update.py (hunk 6 above)
Comment (hanke) [inline datalad/distribution/update.py]: So let's get rid of this one.

File: datalad/distribution/dataset.py (hunk 4 above)
Comment (yarikoptic) [inline datalad/distribution/dataset.py]: yes please ;)

**Comment by bpoldrack** [comment]
I will integrate RF'ing (AnnexRepo.merge/unlock and the helper to determine dataset for the API functions) into the RF PRs, since this isn't urgent regarding the demo, but a matter to get things ironed out. Agree?

**Comment by hanke** [comment]
Not entirely sure if I got it, but I think: Yes!

**Comment by yarikoptic** [comment]
travis failed 1 test in 1 scenario (no ssh if I got it right):
```
======================================================================
FAIL: datalad.interface.tests.test_unlock.test_unlock
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/datalad/datalad/datalad/tests/utils.py", line 697, in newfunc
    t(*(arg + (uri,)), **kw)
  File "/home/travis/build/datalad/datalad/datalad/interface/tests/test_unlock.py", line 91, in test_unlock
    assert_raises(IOError, open, opj(path, 'test-annex.dat'), "w")
AssertionError: IOError not raised
```

**Comment by bpoldrack** [comment]
Yep. Trying to figure it out right now. I test for a file being locked by
annex by trying to write to it and expect an IOError. This fails when the
tests are running with sudo. Need to find another method to confirm the
file is locked, I guess.


2016-06-14 18:32 GMT+02:00 Yaroslav Halchenko <notifications@github.com>:

File: datalad/interface/save.py [hunk 14]
```diff
@@ -6,7 +6,7 @@
 #   copyright and license terms.
 #
 # ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
-"""For now just a wrapper for Dataset.remember_state()
+"""For now just a wrapper for Dataset.save()
```
Comment (hanke) [inline datalad/interface/save.py]: Minor, but this IS Dataset.save() ;-)

**Comment by hanke** [comment]
There are a few things left open from the comments, but I will merge this now (speed/accuracy tradeoff). We can pick it up from here in another PR.

**Pull request merged**